File: py/p4.py
def is_palindrome(x: int) -> bool:
    x = abs(x)

    k = 1
    y = x // 10
    while y > 0:
        k += 1
        y = y // 10

    if k <= 3:
        return x // (10 ** (k - 1)) == x % 10

    for i in range(1, k // 2 + 1):
        h = x // (10 ** (k - i)) % 10
        l = x // (10 ** (i - 1)) % 10
        if h != l:
            return False

    return True

palindrome = 0
for i in range(999, 900, -1):
    for j in range(i, 900, -1):
        prod = i * j
        if is_palindrome(prod) and prod > palindrome:
            palindrome = prod

print(palindrome)

File: hs/p4.hs
base :: Integer
base = 10

thresh :: Int
thresh = 100

numDigits' :: Int -> Integer -> Int
numDigits' acc x =
    let y = abs x in
        if x < base
            then acc
            else numDigits' (acc + 1) (div x base)

numDigits :: Integer -> Int
numDigits = numDigits' 1

isPalindromicNum :: Integer -> Bool
isPalindromicNum x =
    let k = numDigits x in
        if k <= 3
            then let e = base ^ (k - 1) in div x e == mod x base
            else isPalindrome (show x)

isPalindrome :: String -> Bool
isPalindrome [] = True
isPalindrome [x] = True
isPalindrome xs = head xs == last xs && isPalindrome (init (tail xs))

findMaxPalindromicNum :: Integer -> Integer -> Integer -> Integer
findMaxPalindromicNum i j max =
    let prod = i * j
        newMax = if isPalindromicNum prod && prod > max then prod else max
    in
        if j > 900
            then findMaxPalindromicNum i (j - 1) newMax
            else if i > 900
                then findMaxPalindromicNum (i - 1) (i - 1) newMax
                else newMax

main = print $ findMaxPalindromicNum 999 999 0
